### PR TITLE
Routine to remove noise/interference from FITACF files

### PR DESCRIPTION
**Pull request by ecbland**
This PR performs median filtering on a `fitacf` file to remove salt & pepper noise and streaks caused by interference (see #389). 

The filtering is performed using a 3x3 grid in range-time space. If the median of the qflgs in the 3x3 grid is zero, the center record is removed. The filtering is done separately for each beam and channel. The output file is another `fitacf` file. 

**Note:** I wanted to name this new routine `fitfilter`, but that name was [taken](https://github.com/SuperDARN/rst/tree/master/codebase/superdarn/src.bin/tk/tool/fitfilter.1.4) :stuck_out_tongue_winking_eye: . For now it's called `fit_median_filter`, but other suggestions are welcome.
(EDIT: I renamed the function to `fit_speck_removal`, so I've updated the instructions below to match)

------

To test:
```
git checkout EHN/fit-median-filtering
make.build && make.code

fit_speck_removal -vb 20170825.gbr.fitacf3 > filter.fitacf3

# Make range-time plots to compare the data before and after filtering
time_plot -png -a -b 8 -st 00:00 -ex 24:00 20170825.gbr.fitacf3 > original.png
time_plot -png -a -b 8 -st 00:00 -ex 24:00 filter.fitacf3 > filtered.png
```

### CODE REVIEW

**Comment by pasha-ponomarenko** [comment]
@ecbland, thank you very much! I am not sure how I can review this. I guess I will need ot install this branch and to run the code on a FITACF3 file with interference. 
On the name of the routine: I believe _**fit_median_filter**_ is a bit misleading as it alleges the conventional median filtering/smoothing. I think somehting like _**fit_spike_removal**_ , _**fit_speck_removal**_ or _**fit_despecking/despiking**_ would be less confusing.

**Comment by ecbland** [comment]
> @pasha-ponomarenko  I am not sure how I can review this. I guess I will need to install this branch and to run the code on a FITACF3 file with interference.

Yes, switch to this branch and then run the routine on some fitacf files (e.g. the examples you gave in #389). Our focus has been on fitacf3 data, but of course this can be used on fitacf2 data as well. 

Testing of this PR can focus on:
1. Checking that my C implementation of the algorithm is correct
2. Identifying any additional command line options that would be helpful. Currently the only option is the usual "verbose" `-vb`. 
3. Suggesting improvements to the code structure, variable names, comments, etc  
(thanks for the alternative name suggestions)

Some additional science testing with more radars and control programs might be a good idea. We already know (from #389) that this routine will overfilter multifrequency data that has not been separated into different channels, and that near-range scatter will be removed. If there's anything else like that, it would be good to identify and document it.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 1]
```diff
@@ -0,0 +1,322 @@
+/* fit_median_filter.c
+   ==========
+   
+Removes salt & pepper noise from a fitacf file using median filtering. 
+Filtering is performed separately for each beam and channel. 
+Output is a fitacf file.
+
+(C) Copyright 2021 E.C.Bland
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:8]: Just double checking U of S or Unis is not needed on this line? Not sure how your U of S contract worked out.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+#include "dmap.h"
+#include "rprm.h"
+#include "fitdata.h"
+
+#include "fitread.h"
+#include "fitwrite.h"
+
+#include "errstr.h"
+#include "hlpstr.h"
+
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:52]: Add inline comment what each number means and why if relevant

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+  }
+
+  if (version==1) {
+    OptionVersion(stdout);
+    exit(0);
+  }
+  
+  FILE *fp=NULL;
+  if (arg==argc) fp=stdin;
+  else fp=fopen(argv[arg],"r");
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:134]:
``` suggestion
  if (arg==argc) 
      fp=stdin;
  else 
      fp=fopen(argv[arg],"r");
```

Identing is a preferred style as it is easier to read

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  if (fp==NULL) {
+    fprintf(stderr,"File not found.\n");
+    exit(-1);
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:137]: Wanna try using errno for the library? 

https://www.tutorialspoint.com/cprogramming/c_error_handling.htm
https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+  time_t ctime;
+  int c,n;
+  char command[128];
+  char tmstr[40];
+  command[0]=0;
+  n=0;
+  for (c=0;c<argc;c++) {
+    n+=strlen(argv[c])+1;
+    if (n>127) break;
+    if (c !=0) strcat(command," ");
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:150]:
``` suggestion
    if (n>127) 
        break;
    if (c !=0) 
        strcat(command," ");
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    strcat(command,argv[c]);
+  }
+  
+  
+  prm=RadarParmMake();
+  fit=FitMake();
+  if (FitFread(fp,prm,fit)==-1) {
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:157]: fpclose or call your free function?

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    fprintf(stderr,"Error reading file\n");
+    exit(-1);
+  }
+
+  // Initial memory allocation to store qflgs
+  qflgData *qflgs=malloc(sizeof(qflgData));
+  qflgs->value = malloc(tmax*maxrng*maxch*maxbm * sizeof(int));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:164]: Check to make sure the memory was allocated by either checking if they equal null or even better if errno is greater than 1 and print the actual value.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+  qflgs->used = 0;
+  qflgs->size = tmax*maxrng*maxch*maxbm;
+  
+  
+  // Read the FITACF file to populate the qflg array
+  int index;
+  int maxindex=-1;
+  int rng,bm,ch;
+  int nrec[maxbm][maxch]; // counts number of records for each beam/channel
+  memset(nrec,0,sizeof(nrec));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:174]: Doesn't memset return a pointer?

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+  
+  do {
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:176]: What is this loop doing? Add a comment please :)

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+  qflgData *qflgs=malloc(sizeof(qflgData));
+  qflgs->value = malloc(tmax*maxrng*maxch*maxbm * sizeof(int));
+  qflgs->used = 0;
+  qflgs->size = tmax*maxrng*maxch*maxbm;
+  
+  
+  // Read the FITACF file to populate the qflg array
+  int index;
+  int maxindex=-1;
+  int rng,bm,ch;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:172]:
``` suggestion
  int beam, channel;
```
You don't need to define for loop counters a head of time :p

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+  int nrec[maxbm][maxch]; // counts number of records for each beam/channel
+  memset(nrec,0,sizeof(nrec));
+  
+  do {
+  
+    bm=prm->bmnum;
+    ch=prm->channel;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:179]:
``` suggestion
    beam = prm->bmnum;
    channel = prm->channel;
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+    for (rng=0;rng<prm->nrang;rng++) {
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:180]:
``` suggestion
    for (int gate=0; gate < prm->nrang; gate++) {
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+      index=get_index(bm,ch,rng,nrec[bm][ch],maxbm,maxch,maxrng);
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:181]:
``` suggestion
      index=get_index(beam, channel, gate, nrec[beam][channel], maxbm, maxch, maxrng);
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+      if (maxindex < index) maxindex=index;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:182]:
``` suggestion
      if (maxindex < index) 
          maxindex=index;
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+      if (qflgs->size < maxindex) {
+        qflgs->size += tmax*maxrng*maxch*maxbm;
+        qflgs->value = realloc(qflgs->value, qflgs->size * sizeof(int));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:185]: Check if the value is not 0 and errno was not set 
https://pubs.opengroup.org/onlinepubs/009695399/functions/realloc.html

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+      }
+      qflgs->value[index] = fit->rng[rng].qflg;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:187]:
``` suggestion
      qflgs->value[index] = fit->rng[gate].qflg;
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+      qflgs->used = maxindex;
+      
+    }
+    nrec[bm][ch]++;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:191]:
``` suggestion
    nrec[beam][channel]++;
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+  } while (FitFread(fp,prm,fit) !=-1);
+  
+  
+  // Do the median filtering
+  //   Since qflg can only be 0 or 1, the median qflg of the 3x3 grid can be 
+  //     calculated using the test ( sum_of_qflgs >= 5 ).
+  //   Replicate padding is used to handle corner/edge cases
+  int sum;
+  int index_list[9];  
+  memset(index_list,-1,sizeof(index_list));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:202]: check return?

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+  
+  
+  //** Read the file again, apply the median filter, and write a new file
+  // rewind the file pointer
+  rewind(fp);
+  if (FitFread(fp,prm,fit)==-1) {
+    fprintf(stderr,"Error reading file\n");
+    exit(-1);
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:210]: Call your free program to free up all your data so far

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+  //     calculated using the test ( sum_of_qflgs >= 5 ).
+  //   Replicate padding is used to handle corner/edge cases
+  int sum;
+  int index_list[9];  
+  memset(index_list,-1,sizeof(index_list));
+  
+  
+  //** Read the file again, apply the median filter, and write a new file
+  // rewind the file pointer
+  rewind(fp);
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:207]: Does rewind return anything?

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+  if (FitFread(fp,prm,fit)==-1) {
+    fprintf(stderr,"Error reading file\n");
+    exit(-1);
+  }
+  int status=0;
+  memset(nrec,0,sizeof(nrec));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:213]: check return

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+  int t[maxbm][maxch];
+  memset(t,0,sizeof(t));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:215]: check return

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+  
+    if (vb) {
+      fprintf(stderr,"%.4d-%.2d-%.2d %.2d:%.2d:%.2d %.2d %.2d\n",
+              prm->time.yr,prm->time.mo,prm->time.dy,
+              prm->time.hr,prm->time.mt,prm->time.sc,
+              prm->channel,prm->bmnum);
+    }
+    
+    bm=prm->bmnum;
+    ch=prm->channel;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:226]:
``` suggestion
    beam=prm->bmnum;
    channel=prm->channel;
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 24]
```diff
@@ -0,0 +1,322 @@
...
+    for (rng=0;rng<prm->nrang;rng++) {
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:227]:
``` suggestion
    for (int rng = 0; rng<prm->nrang; rng++) {
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 25]
```diff
@@ -0,0 +1,322 @@
...
+      
+      if (fit->rng[rng].qflg==1) {
+
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:230]:
```suggestion
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 26]
```diff
@@ -0,0 +1,322 @@
...
+              prm->time.yr,prm->time.mo,prm->time.dy,
+              prm->time.hr,prm->time.mt,prm->time.sc,
+              prm->channel,prm->bmnum);
+    }
+    
+    bm=prm->bmnum;
+    ch=prm->channel;
+    for (rng=0;rng<prm->nrang;rng++) {
+      
+      if (fit->rng[rng].qflg==1) {
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:229]: If you like my name changes maybe a find all and replace would be easier then me keep suggesting :p

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 27]
```diff
@@ -0,0 +1,322 @@
...
+          sum = 2*qflgs->value[index_list[0]] + qflgs->value[index_list[1]] 
+              + 2*qflgs->value[index_list[3]] + qflgs->value[index_list[4]] 
+              + 2*qflgs->value[index_list[6]] + qflgs->value[index_list[7]];
+        
+        // all other cells
+        else sum = qflgs->value[index_list[0]] + qflgs->value[index_list[1]] 
+                 + qflgs->value[index_list[2]] + qflgs->value[index_list[3]] 
+                 + qflgs->value[index_list[4]] + qflgs->value[index_list[5]] 
+                 + qflgs->value[index_list[6]] + qflgs->value[index_list[7]] 
+                 + qflgs->value[index_list[8]];
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:278]:
``` suggestion
        else 
                sum = qflgs->value[index_list[0]] + qflgs->value[index_list[1]] 
                        + qflgs->value[index_list[2]] + qflgs->value[index_list[3]] 
                        + qflgs->value[index_list[4]] + qflgs->value[index_list[5]] 
                        + qflgs->value[index_list[6]] + qflgs->value[index_list[7]] 
                        + qflgs->value[index_list[8]];
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 28]
```diff
@@ -0,0 +1,322 @@
...
+        
+        // Remove record if median=0 (sum of qflgs < 5)
+        if (sum < 5) fit->rng[rng].qflg=0;
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:281]:
```suggestion
        if (sum < 5) 
            fit->rng[rng].qflg=0;
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 29]
```diff
@@ -0,0 +1,322 @@
...
+    prm->origin.code = 1; // file was not generated at radar site
+    ctime = time((time_t) 0);
+    if (RadarParmSetOriginCommand(prm,command) == -1) 
+    {
+        fprintf(stderr,"Error: cannot set Origin Command\n");
+        free_parameters(prm, fit, fp, qflgs);
+        exit(-1);
+    }
+
+    strcpy(tmstr,asctime(gmtime(&ctime)));
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:297]: check return value or errno

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 30]
```diff
@@ -0,0 +1,322 @@
...
+    
+    // write the output file
+    status=FitFwrite(stdout,prm,fit);
+    if (status ==-1) {
+        fprintf(stderr,"Error writing output file.\n");
+        exit(-1);
+    }
+  
+  } while (FitFread(fp,prm,fit) !=-1);
+  if (fp !=stdin) fclose(fp);
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:314]:
``` suggestion

  if (fp !=stdin) 
      fclose(fp);
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 31]
```diff
@@ -0,0 +1,322 @@
...
+  
+
+  // Free memory
+  free_parameters(prm, fit, NULL, qflgs);
+  
+
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:320]:
```suggestion
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 32]
```diff
@@ -0,0 +1,322 @@
...
+    // write the output file
+    status=FitFwrite(stdout,prm,fit);
+    if (status ==-1) {
+        fprintf(stderr,"Error writing output file.\n");
+        exit(-1);
+    }
+  
+  } while (FitFread(fp,prm,fit) !=-1);
+  if (fp !=stdin) fclose(fp);
+
```
Comment (mts299) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:315]:
```suggestion
```

**Review by mts299 — COMMENTED** [review]
Looks good a few suggestions:
- [ ] use indenting, inline if statements can easily be misread so indent 
- [ ] always check return values 
- [ ] use errno to save you on error message. Check out make_grid PR for example on how to use the library 
- [ ] Add comments on each loop as it makes it easier for the user to know what is going on in the body 
- [ ] make sure you avoid 2 letter or small variable names.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 6 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:161]: ok, added below.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c [hunk 33]
```diff
@@ -0,0 +1,322 @@
...
+    if (c !=0) strcat(command," ");
+    strcat(command,argv[c]);
+  }
+  
+  
+  prm=RadarParmMake();
+  fit=FitMake();
+  if (FitFread(fp,prm,fit)==-1) {
+    fprintf(stderr,"Error reading file\n");
+    exit(-1);
```
Comment (ecbland) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:163]:
```suggestion
    free_parameters(prm, fit, fp, NULL);
    exit(-1);
```

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 9 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:180]: There's already a comment on line 169. Is that enough?

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 11 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:183]: I don't mind changing this, but I deliberately used `bm` and `ch` to match the rest of RST. Is it better to be consistent or descriptive?

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 20 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:212]: I don't think so. This is how it's used elsewhere in RST, so I just copied it.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 19 above)
Comment (ecbland) [inline codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:215]:
```suggestion
    free_parameters(prm, fit, fp, qflgs);
    exit(-1);
```

Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:215]: Good catch. I added this on the line above

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 2 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:52]: ok, see below

**Review by ecbland — COMMENTED** [review]
Thanks for the review, @mts299!
I still need to look at the errno and memset return value comments, but I've responded to everything else.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 11 above)
Comment (mts299) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c:185]: descriptive in this case. Consistency is good if consistency was descriptive ;) 

Just like the disclaimer the files have to stand alone to make sense to someone

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 8 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c]: @mts299 I'm not sure what you're asking me to do here. Can you make a suggestion of how I should change this? `memset` is used this way throughout RST (i.e. without checking/using the return value), so I've just followed that.

File: codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c (hunk 29 above)
Comment (ecbland) [reply codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c]: Again, could you please make a suggestion on how I should change this? I just copied this line from elsewhere in RST. Thanks :smile:

**Comment by pasha-ponomarenko** [comment]
@ecbland , I will try to test the code as soon as I get time, probably earlier next week as I go on vacations.

I am still a bit concerned about the routine's name as it might be confused with the conventional media=n filtering/smoothing (see my previous [comments](https://github.com/SuperDARN/rst/pull/400?_pjax=%2523js-repo-pjax-container#issuecomment-799956272))

**Comment by ecbland** [comment]
> I am still a bit concerned about the routine's name as it might be confused with the conventional median filtering/smoothing

@pasha-ponomarenko I renamed the routine 2 days ago, but that probably wasn't obvious! It's now called `fit_speck_removal`

**Comment by pasha-ponomarenko** [comment]
@ecbland Apologies, I have missed it! 
:-(

**Comment by pasha-ponomarenko** [comment]
@ecbland , for starters, here are original and "de-specked" data from RKN, which relatively recently also started to suffer from local interference:


original


filtered

So far it looks good to me but I wil also test HOK/HKW and CVE/CVW pairs.

**Comment by pasha-ponomarenko** [comment]
It might also make sense to plot FITACF2.5 data for comparison.

File: codebase/superdarn/src.bin/tk/tool/fit_speck_removal.1.0/fit_speck_removal.c [hunk 34]
```diff
@@ -0,0 +1,364 @@
...
+        else if (irec[beam][channel]==0)
+          sum = 2*qflgs->value[index_list[0]] + 2*qflgs->value[index_list[1]] 
+              + 2*qflgs->value[index_list[2]] + qflgs->value[index_list[6]] 
+              + qflgs->value[index_list[7]]   + qflgs->value[index_list[8]];
+        else if (irec[beam][channel]==nrec[beam][channel])
+          {sum = 2*qflgs->value[index_list[0]] + 2*qflgs->value[index_list[1]] 
+              + 2*qflgs->value[index_list[2]] + qflgs->value[index_list[3]] 
+              + qflgs->value[index_list[4]]   + qflgs->value[index_list[5]];
+              fprintf(stderr,"%d %d\n",irec[beam][channel],nrec[beam][channel]);
+              }
```
Comment (ecbland) [inline codebase/superdarn/src.bin/tk/tool/fit_speck_removal.1.0/fit_speck_removal.c:299]: Just deleting some test code I forgot about
```suggestion
          sum = 2*qflgs->value[index_list[0]] + 2*qflgs->value[index_list[1]] 
              + 2*qflgs->value[index_list[2]] + qflgs->value[index_list[3]] 
              + qflgs->value[index_list[4]]   + qflgs->value[index_list[5]];
```

**Comment by pasha-ponomarenko** [comment]
@ecbland, here are some tests for HOK:

FITACF3:

image

FITACF3 despecked:

image

FITACF2:

image


I used my own IDL plotting routine as there is somehting funny with _time_plot.c_  
It seems to economise on resolution (there are only ~60 time bins for a 3-hour interval), and pixels from adjacent time stamps sometimes overlap. As a result, the range-time plots look much more densely populated by echoes. Here are the same data as above for FITACF3 (no speck removal applied) but plotted by _time_plot_:

20200220 C0 hok fitacf3

As one can see there apparently very few blank pixels.

If I rmember correctly,  similar concerns hav been voiced before but cannot recall if this issue was reloved.

**Comment by pasha-ponomarenko** [comment]
Same for CVE with verical stripes:

FITACF3:

image


FITACF3 despecked:

image


FITACF2:

image

**Comment by ecbland** [comment]
@pasha-ponomarenko Thanks for testing! It seems to be behaving as expected :smile: 

> I used my own IDL plotting routine as there is somehting funny with time_plot.c
> It seems to economise on resolution (there are only ~60 time bins for a 3-hour interval), and pixels from adjacent time stamps sometimes overlap. As a result, the range-time plots look much more densely populated by echoes. 

Yes, `time_plot` is a bit primitive for this application so it's best to use your own code for plotting (which you've done)

**Comment by pasha-ponomarenko** [comment]
In summary, the routine works as intended, including two-frequency mode (I tested it on some RKN data as well). 
If a single pass of despecking is not 100% saisfactory, it can be run again. :-)

**Review by pasha-ponomarenko — COMMENTED** [review]
In summary, the routine works as intended, including two-frequency mode (I tested it on some RKN data as well).
If a single pass of despecking is not 100% saisfactory, it can be run again. :-)

**Comment by ecbland** [comment]
This PR has been tested/reviewed by @pasha-ponomarenko and @mts299 -- is it ready to merge now?

**Comment by pasha-ponomarenko** [comment]
@ecbland , as far as I am concerned -- yes, but I have a vested interest here... 
;-)

**Comment by pasha-ponomarenko** [comment]
@ecbland , keeping in mind your new job, I wonder if you have time and desire to instruct on how to create a similar script for I&Q data? It seems like we have found an algorithmic solution for HOK/HKW cross-interference which allows to clear the historic data as well. It is simpler in comparison with the de-spiking of FITACF data as it requires only samples from a single pulse sequence.
